dspace-rest/src/main/java/org/dspace/rest/HierarchyResource.java

```diff
@@ -0,0 +1,134 @@
+import org.dspace.rest.common.HierarchyRepository;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.*;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.HttpHeaders;
+
+import java.io.UnsupportedEncodingException;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+
+/*
+ * This class retrieves the community hierarchy in an optimized format.
+ * 
+ * @author Terry Brady, Georgetown University
+ */
+@Path("/hierarchy")
+@Produces({ MediaType.APPLICATION_JSON, MediaType.APPLICATION_XML })
+public class HierarchyResource extends Resource {
+    private static Logger log = Logger.getLogger(HierarchyResource.class);
+    protected SiteService siteService = ContentServiceFactory.getInstance().getSiteService();
+    protected CommunityService communityService = ContentServiceFactory.getInstance().getCommunityService();
+    protected AuthorizeService authorizeService = AuthorizeServiceFactory.getInstance().getAuthorizeService();
+
+    /**
+    * @param headers
+     *            If you want to access to collection under logged user into
+     *            context. In headers must be set header "rest-dspace-token"
+     *            with passed token from login method.
+     * @return Return instance of collection. It can also return status code
+     *         NOT_FOUND(404) if id of collection is incorrect or status code
+     * @throws UnsupportedEncodingException 
+     * @throws WebApplicationException
+     *             It is thrown when was problem with database reading
+     *             (SQLException) or problem with creating
+     *             context(ContextException). It is thrown by NOT_FOUND and
+     *             UNATHORIZED status codes, too.
+     */
+	@GET
+    @Produces({MediaType.APPLICATION_XML})
+    public HierarchyRepository getHierarchy(
+    		@QueryParam("userAgent") String user_agent, @QueryParam("xforwarderfor") String xforwarderfor,
+    		@Context HttpHeaders headers, @Context HttpServletRequest request) throws UnsupportedEncodingException {
+		
+		org.dspace.core.Context context = null;
+		HierarchyRepository repo = new HierarchyRepository();
+		
+        try {
+            context = createContext(getUser(headers));
+            if (ConfigurationManager.getBooleanProperty("rest", "rest-hierarchy-authenticate", true) == false) {
+                context.turnOffAuthorisationSystem();            	
+            }
+
+            Site site = siteService.findSite(context);
+            repo.setId(site.getID().toString());
+            repo.setName("Repository");
+            repo.setHandle(site.getHandle());
+    		List<Community> dspaceCommunities = communityService.findAllTop(context);
+    		processCommunity(context, repo, dspaceCommunities);
+
+    		return repo;
+        } catch (Exception e) {
+        	e.printStackTrace();
+            log.error(e.getMessage());
```

Comment:
Best to remove the "e.printStackTrace()" & log the entire error, not just the message.